File: doa/nodes/agent_signal_fusion.py
"""Agent signal fusion node for LangGraph workflow."""

import logging
import time
from typing import Any, Dict, List
import statistics

from models.state import GraphState
from models.types import AuditEntry, AgentSignal, FusedSignal
from config import EngineConfig

logger = logging.getLogger(__name__)


def calculate_weighted_probability(
    signals: List[AgentSignal],
    config: EngineConfig
) -> float:
    """
    Calculate weighted average probability from agent signals.
    
    Weighting Strategy:
    1. Base weight = agent confidence (0-1)
    2. Historical accuracy weight (if available in metadata)
    3. Normalize weights to sum to 1.0
    4. Calculate weighted average of fair_probability values
    
    Args:
        signals: List of agent signals
        config: Engine configuration
        
    Returns:
        Weighted probability (0-1)
        
    Examples:
        >>> signals = [
        ...     AgentSignal(confidence=0.8, fair_probability=0.6, ...),
        ...     AgentSignal(confidence=0.6, fair_probability=0.7, ...)
        ... ]
        >>> prob = calculate_weighted_probability(signals, config)
        >>> assert 0.0 <= prob <= 1.0
    """
    if not signals:
        return 0.5  # Default neutral probability
    
    # Calculate weights for each signal
    weights = []
    probabilities = []
    
    for signal in signals:
        # Base weight from confidence
        weight = signal.confidence
        
        # Adjust weight based on historical accuracy if available
        historical_accuracy = signal.metadata.get('historical_accuracy')
        if historical_accuracy is not None:
            # Boost weight for agents with good track record
            weight *= (0.5 + 0.5 * historical_accuracy)
        
        weights.append(weight)
        probabilities.append(signal.fair_probability)
    
    # Normalize weights
    total_weight = sum(weights)
    if total_weight == 0:
        # All zero confidence - use equal weighting
        normalized_weights = [1.0 / len(weights)] * len(weights)
    else:
        normalized_weights = [w / total_weight for w in weights]
    
    # Calculate weighted average
    weighted_prob = sum(
        prob * weight 
        for prob, weight in zip(probabilities, normalized_weights)
    )
    
    return weighted_prob


def calculate_signal_alignment(signals: List[AgentSignal]) -> float:
    """
    Calculate how aligned agent signals are.
    
    Alignment is measured by:
    1. Standard deviation of fair_probability values (lower = more aligned)
    2. Consistency of direction (YES/NO/NEUTRAL)
    3. Normalized to 0-1 scale (1 = perfect alignment, 0 = maximum disagreement)
    
    Args:
        signals: List of agent signals
        
    Returns:
        Alignment score (0-1)
        
    Examples:
        >>> signals = [
        ...     AgentSignal(direction="YES", fair_probability=0.6, ...),
        ...     AgentSignal(direction="YES", fair_probability=0.62, ...)
        ... ]
        >>> alignment = calculate_signal_alignment(signals)
        >>> assert alignment > 0.8  # High alignment
    """
    if len(signals) < 2:
        return 1.0  # Single signal is perfectly aligned with itself
    
    # Calculate probability standard deviation
    probabilities = [s.fair_probability for s in signals]
    prob_std = statistics.stdev(probabilities)
    
    # Normalize std to 0-1 scale (0.5 std = maximum disagreement)
    # Lower std = higher alignment
    prob_alignment = max(0.0, 1.0 - (prob_std / 0.5))
    
    # Calculate direction consistency
    directions = [s.direction for s in signals]
    direction_counts = {
        'YES': directions.count('YES'),
        'NO': directions.count('NO'),
        'NEUTRAL': directions.count('NEUTRAL')
    }
    
    # Direction alignment = proportion of most common direction
    max_direction_count = max(direction_counts.values())
    direction_alignment = max_direction_count / len(signals)
    
    # Combined alignment (weighted average)
    alignment = 0.6 * prob_alignment + 0.4 * direction_alignment
    
    return alignment


def detect_conflicts(signals: List[AgentSignal]) -> List[str]:
    """
    Detect conflicting signals between agents.
    
    Conflicts are identified when:
    1. Agents have opposite directions (YES vs NO) with high confidence
    2. Agents have significantly different fair_probability estimates (>0.3 difference)
    3. Agents identify contradictory key drivers
    
    Args:
        signals: List of agent signals
        
    Returns:
        List of conflict descriptions
        
    Examples:
        >>> signals = [
        ...     AgentSignal(agent_name="A", direction="YES", confidence=0.8, ...),
        ...     AgentSignal(agent_name="B", direction="NO", confidence=0.8, ...)
        ... ]
        >>> conflicts = detect_conflicts(signals)
        >>> assert len(conflicts) > 0
    """
    conflicts = []
    
    # Check for directional conflicts
    yes_signals = [s for s in signals if s.direction == 'YES' and s.confidence > 0.7]
    no_signals = [s for s in signals if s.direction == 'NO' and s.confidence > 0.7]
    
    if yes_signals and no_signals:
        yes_agents = [s.agent_name for s in yes_signals]
        no_agents = [s.agent_name for s in no_signals]
        conflicts.append(
            f"Directional conflict: {', '.join(yes_agents)} favor YES "
            f"while {', '.join(no_agents)} favor NO"
        )
    
    # Check for probability conflicts (>0.3 difference)
    for i, signal_a in enumerate(signals):
        for signal_b in signals[i+1:]:
            prob_diff = abs(signal_a.fair_probability - signal_b.fair_probability)
            if prob_diff > 0.3:
                conflicts.append(
                    f"Probability conflict: {signal_a.agent_name} estimates "
                    f"{signal_a.fair_probability:.2f} while {signal_b.agent_name} "
                    f"estimates {signal_b.fair_probability:.2f}"
                )
    
    return conflicts


async def agent_signal_fusion_node(
    state: GraphState,
    config: EngineConfig
) -> Dict[str, Any]:
    """
    Aggregate agent signals with dynamic weighting.
    
    This node is the fan-in point after parallel agent execution. It:
    1. Waits for all parallel agents to complete (handled by LangGraph)
    2. Aggregates agent signals using confidence-based weighting
    3. Calculates signal alignment to measure consensus
    4. Detects conflicts between agent signals
    5. Produces a fused signal for downstream nodes
    
    Weighting Strategy:
    - Signals are weighted by agent confidence (0-1)
    - Historical accuracy (if available) adjusts weights
    - Weights are normalized to sum to 1.0
    
    Args:
        state: Current workflow state with agent_signals
        config: Engine configuration
        
    Returns:
        State update with fused_signal and audit entry
        
    State Requirements:
        - agent_signals: List of AgentSignal from parallel agents (required)
        - active_agents: List of agent names that were dispatched (optional)
        
    State Updates:
        - fused_signal: FusedSignal with aggregated analysis
        - audit_log: Audit entry for signal fusion stage
        
    Examples:
        >>> state = {
        ...     "agent_signals": [
        ...         AgentSignal(agent_name="A", confidence=0.8, fair_probability=0.6, ...),
        ...         AgentSignal(agent_name="B", confidence=0.7, fair_probability=0.65, ...)
        ...     ]
        ... }
        >>> result = await agent_signal_fusion_node(state, config)
        >>> assert "fused_signal" in result
    """
    start_time = time.time()
    
    # Extract agent signals from state
    agent_signals_raw = state.get("agent_signals", [])
    active_agents = state.get("active_agents", [])
    
    # Convert dictionaries to AgentSignal objects if needed
    agent_signals = []
    for signal in agent_signals_raw:
        if isinstance(signal, dict):
            # Convert dict to AgentSignal
            agent_signals.append(AgentSignal(**signal))
        else:
            # Already an AgentSignal object
            agent_signals.append(signal)
    
    # Validate we have signals
    if not agent_signals:
        logger.warning("Signal fusion node called with no agent signals")
        
        # Return neutral fused signal
        return {
            "fused_signal": FusedSignal(
                weighted_probability=0.5,
                signal_alignment=0.0,
                conflicts=["No agent signals available"],
                contributing_agents=[]
            ),
            "audit_log": [AuditEntry(
                stage="agent_signal_fusion",
                timestamp=int(time.time()),
                status="completed",
                details={
                    "warning": "No agent signals to fuse",
                    "signal_count": 0
                }
            )]
        }
    
    logger.info(f"Fusing {len(agent_signals)} agent signals")
    
    try:
        # Calculate weighted probability
        weighted_probability = calculate_weighted_probability(agent_signals, config)
        
        # Calculate signal alignment
        signal_alignment = calculate_signal_alignment(agent_signals)
        
        # Detect conflicts
        conflicts = detect_conflicts(agent_signals)
        
        # Get contributing agent names
        contributing_agents = [s.agent_name for s in agent_signals]
        
        # Create fused signal
        fused_signal = FusedSignal(
            weighted_probability=weighted_probability,
            signal_alignment=signal_alignment,
            conflicts=conflicts,
            contributing_agents=contributing_agents
        )
        
        duration_ms = int((time.time() - start_time) * 1000)
        
        # Calculate statistics for logging
        avg_confidence = statistics.mean([s.confidence for s in agent_signals])
        prob_std = statistics.stdev([s.fair_probability for s in agent_signals]) if len(agent_signals) > 1 else 0.0
        
        logger.info(
            f"Signal fusion completed in {duration_ms}ms: "
            f"weighted_prob={weighted_probability:.3f}, "
            f"alignment={signal_alignment:.3f}, "
            f"conflicts={len(conflicts)}"
        )
        
        return {
            "fused_signal": fused_signal,
            "audit_log": [AuditEntry(
                stage="agent_signal_fusion",
                timestamp=int(time.time()),
                status="completed",
                details={
                    "duration_ms": duration_ms,
                    "signal_count": len(agent_signals),
                    "weighted_probability": weighted_probability,
                    "signal_alignment": signal_alignment,
                    "conflict_count": len(conflicts),
                    "contributing_agents": contributing_agents,
                    "avg_confidence": avg_confidence,
                    "probability_std": prob_std,
                    "expected_agents": len(active_agents),
                    "received_agents": len(agent_signals)
                }
            )]
        }
    
    except Exception as e:
        duration_ms = int((time.time() - start_time) * 1000)
        logger.error(f"Signal fusion failed after {duration_ms}ms: {e}")
        
        # Return fallback fused signal
        return {
            "fused_signal": FusedSignal(
                weighted_probability=0.5,
                signal_alignment=0.0,
                conflicts=[f"Fusion error: {str(e)}"],
                contributing_agents=[]
            ),
            "audit_log": [AuditEntry(
                stage="agent_signal_fusion",
                timestamp=int(time.time()),
                status="failed",
                details={
                    "duration_ms": duration_ms,
                    "error": str(e),
                    "signal_count": len(agent_signals)
                }
            )]
        }


def create_agent_signal_fusion_node(config: EngineConfig):
    """
    Factory function to create agent signal fusion node with dependencies.
    
    This factory pattern allows the node to be created with the required
    dependencies (config) while maintaining the standard LangGraph node signature.
    
    Args:
        config: Engine configuration
        
    Returns:
        Async function that takes state and returns state update
        
    Examples:
        >>> config = load_config()
        >>> node = create_agent_signal_fusion_node(config)
        >>> result = await node(state)
    """
    async def node(state: GraphState) -> Dict[str, Any]:
        return await agent_signal_fusion_node(state, config)
    
    return node

File: tradewizard-agents/src/nodes/agent-signal-fusion.ts
/**
 * Agent Signal Fusion Node
 *
 * This LangGraph node aggregates and weights signals from all active agents,
 * detects conflicts, calculates alignment, and produces a unified fused signal.
 *
 * The fusion process:
 * 1. Groups signals by agent type
 * 2. Applies dynamic weighting based on agent type and market context
 * 3. Detects signal conflicts (divergent probability estimates)
 * 4. Calculates signal alignment (how much agents agree)
 * 5. Computes fusion confidence with alignment bonus and quality penalty
 * 6. Produces a FusedSignal with weighted consensus probability
 */

import type { GraphStateType } from '../models/state.js';
import type { AgentSignal } from '../models/types.js';
import type { EngineConfig } from '../config/index.js';
import { getPerformanceWeightAdjustment } from '../utils/performance-tracking.js';

/**
 * Agent type classification for weighting
 */
type AgentType =
  | 'mvp'
  | 'event_intelligence'
  | 'polling_statistical'
  | 'sentiment_narrative'
  | 'price_action'
  | 'event_scenario';

/**
 * Classify agent by type based on agent name
 *
 * @param agentName - Name of the agent
 * @returns Agent type classification
 */
function classifyAgentType(agentName: string): AgentType {
  // MVP agents
  if (
    ['market_microstructure', 'probability_baseline', 'risk_assessment'].includes(agentName)
  ) {
    return 'mvp';
  }

  // Event intelligence agents
  if (['breaking_news', 'event_impact'].includes(agentName)) {
    return 'event_intelligence';
  }

  // Polling & statistical agents
  if (['polling_intelligence', 'historical_pattern'].includes(agentName)) {
    return 'polling_statistical';
  }

  // Sentiment & narrative agents
  if (['media_sentiment', 'social_sentiment', 'narrative_velocity'].includes(agentName)) {
    return 'sentiment_narrative';
  }

  // Price action agents
  if (['momentum', 'mean_reversion'].includes(agentName)) {
    return 'price_action';
  }

  // Event scenario agents
  if (['catalyst', 'tail_risk'].includes(agentName)) {
    return 'event_scenario';
  }

  // Default to MVP for unknown agents
  return 'mvp';
}

/**
 * Get base weight for agent type
 *
 * Base weights reflect the general reliability and importance of each agent type:
 * - MVP agents: 1.0x (baseline)
 * - Event intelligence: 1.2x (high value for event-driven markets)
 * - Polling: 1.5x (very reliable for election markets)
 * - Sentiment: 0.8x (noisy but useful)
 * - Price action: 1.0x (reliable for liquid markets)
 * - Event scenario: 1.0x (forward-looking analysis)
 *
 * @param agentType - Agent type classification
 * @param config - Engine configuration
 * @returns Base weight multiplier
 */
function getBaseWeight(agentType: AgentType, config: EngineConfig): number {
  const baseWeights = config.signalFusion.baseWeights;

  // Use configured base weights if available
  if (baseWeights && baseWeights[agentType] !== undefined) {
    return baseWeights[agentType];
  }

  // Default base weights
  switch (agentType) {
    case 'mvp':
      return 1.0;
    case 'event_intelligence':
      return 1.2;
    case 'polling_statistical':
      return 1.5;
    case 'sentiment_narrative':
      return 0.8;
    case 'price_action':
      return 1.0;
    case 'event_scenario':
      return 1.0;
    default:
      return 1.0;
  }
}

/**
 * Apply context adjustments to base weight
 *
 * Context adjustments modify weights based on:
 * - Agent confidence (higher confidence = higher weight)
 * - Data freshness (stale data = lower weight)
 * - Market liquidity (for price action agents)
 * - Historical performance (high accuracy = higher weight)
 *
 * @param signal - Agent signal
 * @param baseWeight - Base weight for agent type
 * @param state - Current graph state
 * @param config - Engine configuration
 * @returns Adjusted weight
 */
function applyContextAdjustments(
  signal: AgentSignal,
  baseWeight: number,
  state: GraphStateType,
  config: EngineConfig
): number {
  if (!config.signalFusion.contextAdjustments) {
    return baseWeight;
  }

  let adjustedWeight = baseWeight;

  // Confidence adjustment: scale by agent confidence
  // High confidence (>0.8) increases weight by up to 20%
  // Low confidence (<0.5) decreases weight by up to 30%
  const confidenceMultiplier = 0.7 + signal.confidence * 0.5;
  adjustedWeight *= confidenceMultiplier;

  // Data freshness adjustment
  if (state.externalData?.dataFreshness) {
    const agentType = classifyAgentType(signal.agentName);
    let dataSource: string | null = null;

    // Map agent type to data source
    if (agentType === 'event_intelligence' || agentType === 'sentiment_narrative') {
      dataSource = 'news';
    } else if (agentType === 'polling_statistical') {
      dataSource = 'polling';
    }

    // Check data freshness
    if (dataSource && state.externalData.dataFreshness[dataSource]) {
      const dataAge = Date.now() - state.externalData.dataFreshness[dataSource];
      const maxAge = 3600000; // 1 hour in milliseconds

      // If data is stale (>1 hour old), reduce weight
      if (dataAge > maxAge) {
        const staleness = Math.min(dataAge / maxAge, 3); // Cap at 3x max age
        const freshnessMultiplier = 1 / (1 + staleness * 0.2); // Up to 40% reduction
        adjustedWeight *= freshnessMultiplier;
      }
    }
  }

  // Liquidity adjustment for price action agents
  if (
    ['momentum', 'mean_reversion'].includes(signal.agentName) &&
    state.mbd
  ) {
    const liquidityScore = state.mbd.liquidityScore;
    // Low liquidity (<5) reduces price action agent weight
    if (liquidityScore < 5) {
      const liquidityMultiplier = 0.5 + liquidityScore * 0.1; // 0.5x to 1.0x
      adjustedWeight *= liquidityMultiplier;
    }
  }

  // Performance-based adjustment: scale by historical accuracy
  // High accuracy agents (>0.7) get up to 1.5x weight
  // Low accuracy agents (<0.4) get down to 0.5x weight
  const performanceMultiplier = getPerformanceWeightAdjustment(
    signal.agentName,
    state.agentPerformance,
    config
  );
  adjustedWeight *= performanceMultiplier;

  return adjustedWeight;
}

/**
 * Calculate dynamic weights for all agent signals
 *
 * Weights are calculated based on:
 * 1. Base weight for agent type
 * 2. Context adjustments (confidence, data freshness, liquidity)
 * 3. Normalization to sum to 1.0
 *
 * Error handling:
 * - Falls back to equal weights if calculation fails
 * - Ensures all weights are non-negative
 * - Ensures weights sum to 1.0
 *
 * @param signals - Array of agent signals
 * @param state - Current graph state
 * @param config - Engine configuration
 * @returns Record mapping agent name to normalized weight
 */
function calculateDynamicWeights(
  signals: AgentSignal[],
  state: GraphStateType,
  config: EngineConfig
): Record<string, number> {
  try {
    const weights: Record<string, number> = {};

    // Calculate raw weights for each agent
    for (const signal of signals) {
      const agentType = classifyAgentType(signal.agentName);
      const baseWeight = getBaseWeight(agentType, config);
      const adjustedWeight = applyContextAdjustments(signal, baseWeight, state, config);
      weights[signal.agentName] = Math.max(0, adjustedWeight); // Ensure non-negative
    }

    // Normalize weights to sum to 1.0
    const totalWeight = Object.values(weights).reduce((sum, w) => sum + w, 0);

    if (totalWeight > 0) {
      for (const agentName in weights) {
        weights[agentName] = weights[agentName] / totalWeight;
      }
    } else {
      // If all weights are zero, use equal weights
      console.warn('[SignalFusion] All weights are zero, falling back to equal weights');
      const equalWeight = 1.0 / signals.length;
      for (const signal of signals) {
        weights[signal.agentName] = equalWeight;
      }
    }

    return weights;
  } catch (error) {
    // If weight calculation fails, fall back to equal weights
    console.error(
      '[SignalFusion] Weight calculation failed, falling back to equal weights:',
      error instanceof Error ? error.message : String(error)
    );

    const equalWeight = 1.0 / signals.length;
    const weights: Record<string, number> = {};
    for (const signal of signals) {
      weights[signal.agentName] = equalWeight;
    }
    return weights;
  }
}

/**
 * Calculate weighted consensus probability
 *
 * @param signals - Array of agent signals
 * @param weights - Weight for each agent
 * @returns Weighted average probability
 */
function calculateWeightedProbability(
  signals: AgentSignal[],
  weights: Record<string, number>
): number {
  let weightedSum = 0;

  for (const signal of signals) {
    const weight = weights[signal.agentName] || 0;
    weightedSum += signal.fairProbability * weight;
  }

  // Clamp to [0, 1]
  return Math.max(0, Math.min(1, weightedSum));
}

/**
 * Identify conflicting signals
 *
 * Two signals are considered conflicting if their fair probability estimates
 * differ by more than the configured threshold (default: 0.20).
 *
 * @param signals - Array of agent signals
 * @param config - Engine configuration
 * @returns Array of conflicting signal pairs
 */
function identifyConflicts(
  signals: AgentSignal[],
  config: EngineConfig
): Array<{ agent1: string; agent2: string; disagreement: number }> {
  const conflicts: Array<{ agent1: string; agent2: string; disagreement: number }> = [];
  const threshold = config.signalFusion.conflictThreshold;

  // Compare all pairs of signals
  for (let i = 0; i < signals.length; i++) {
    for (let j = i + 1; j < signals.length; j++) {
      const signal1 = signals[i];
      const signal2 = signals[j];

      const disagreement = Math.abs(signal1.fairProbability - signal2.fairProbability);

      if (disagreement > threshold) {
        conflicts.push({
          agent1: signal1.agentName,
          agent2: signal2.agentName,
          disagreement,
        });
      }
    }
  }

  return conflicts;
}

/**
 * Calculate signal alignment
 *
 * Alignment measures how much agents agree with each other.
 * It's calculated as 1 - (standard deviation of probabilities).
 *
 * High alignment (close to 1.0) means agents mostly agree.
 * Low alignment (close to 0.0) means agents strongly disagree.
 *
 * @param signals - Array of agent signals
 * @returns Alignment score (0-1)
 */
function calculateSignalAlignment(signals: AgentSignal[]): number {
  if (signals.length === 0) return 1.0;
  if (signals.length === 1) return 1.0;

  const probabilities = signals.map((s) => s.fairProbability);

  // Calculate mean
  const mean = probabilities.reduce((sum, p) => sum + p, 0) / probabilities.length;

  // Calculate standard deviation
  const variance =
    probabilities.reduce((sum, p) => sum + Math.pow(p - mean, 2), 0) / probabilities.length;
  const stdDev = Math.sqrt(variance);

  // Alignment is inverse of standard deviation
  // Standard deviation ranges from 0 (perfect agreement) to ~0.5 (maximum disagreement)
  // We map this to alignment: 1.0 (perfect) to 0.0 (maximum disagreement)
  const alignment = Math.max(0, 1 - stdDev * 2);

  return alignment;
}

/**
 * Assess data quality across all signals
 *
 * Data quality is based on:
 * - Data freshness (from external data sources)
 * - Agent confidence levels
 *
 * @param signals - Array of agent signals
 * @param state - Current graph state
 * @returns Data quality score (0-1)
 */
function assessDataQuality(signals: AgentSignal[], state: GraphStateType): number {
  let qualityScore = 0;
  let factorCount = 0;

  // Factor 1: Average agent confidence
  const avgConfidence =
    signals.reduce((sum, s) => sum + s.confidence, 0) / signals.length;
  qualityScore += avgConfidence;
  factorCount++;

  // Factor 2: Data freshness
  if (state.externalData?.dataFreshness) {
    const freshness = state.externalData.dataFreshness;
    const now = Date.now();
    const maxAge = 3600000; // 1 hour

    let freshnessScore = 0;
    let sourceCount = 0;

    for (const source in freshness) {
      const age = now - freshness[source];
      const sourceFreshness = Math.max(0, 1 - age / maxAge);
      freshnessScore += sourceFreshness;
      sourceCount++;
    }

    if (sourceCount > 0) {
      qualityScore += freshnessScore / sourceCount;
      factorCount++;
    }
  }

  return factorCount > 0 ? qualityScore / factorCount : 0.5;
}

/**
 * Calculate fusion confidence
 *
 * Fusion confidence is calculated as:
 * - Base confidence: weighted average of agent confidences
 * - Alignment bonus: increase confidence if agents agree (up to 20%)
 * - Data quality penalty: decrease confidence if data is stale (up to 30%)
 *
 * @param signals - Array of agent signals
 * @param weights - Weight for each agent
 * @param alignment - Signal alignment score
 * @param dataQuality - Data quality score
 * @param config - Engine configuration
 * @returns Fusion confidence (0-1)
 */
function calculateFusionConfidence(
  signals: AgentSignal[],
  weights: Record<string, number>,
  alignment: number,
  dataQuality: number,
  config: EngineConfig
): number {
  // Base confidence: weighted average of agent confidences
  let baseConfidence = 0;
  for (const signal of signals) {
    const weight = weights[signal.agentName] || 0;
    baseConfidence += signal.confidence * weight;
  }

  // Alignment bonus: up to 20% increase for high alignment
  const alignmentBonus = alignment * config.signalFusion.alignmentBonus;

  // Data quality penalty: up to 30% decrease for low quality
  const qualityPenalty = (1 - dataQuality) * 0.3;

  // Calculate final confidence
  const confidence = baseConfidence + alignmentBonus - qualityPenalty;

  // Clamp to [0, 1]
  return Math.max(0, Math.min(1, confidence));
}

/**
 * Agent Signal Fusion Node
 *
 * Aggregates and weights signals from all active agents to produce a unified fused signal.
 *
 * @param state - Current graph state
 * @param config - Engine configuration
 * @returns Partial state update with fusedSignal and audit log
 */
export async function agentSignalFusionNode(
  state: GraphStateType,
  config: EngineConfig
): Promise<Partial<GraphStateType>> {
  const startTime = Date.now();
  const { agentSignals, activeAgents } = state;

  // Check if we have agent signals
  if (!agentSignals || agentSignals.length === 0) {
    return {
      fusedSignal: null,
      auditLog: [
        {
          stage: 'agent_signal_fusion',
          timestamp: Date.now(),
          data: {
            success: false,
            reason: 'No agent signals available for fusion',
            duration: Date.now() - startTime,
          },
        },
      ],
    };
  }

  try {
    // Step 1: Calculate dynamic weights
    const weights = calculateDynamicWeights(agentSignals, state, config);

    // Step 2: Calculate weighted consensus probability
    const fairProbability = calculateWeightedProbability(agentSignals, weights);

    // Step 3: Identify conflicting signals
    const conflictingSignals = identifyConflicts(agentSignals, config);

    // Step 4: Calculate signal alignment
    const signalAlignment = calculateSignalAlignment(agentSignals);

    // Check for extreme divergence (signals span full probability range)
    const probabilities = agentSignals.map((s) => s.fairProbability);
    const minProb = Math.min(...probabilities);
    const maxProb = Math.max(...probabilities);
    const probabilityRange = maxProb - minProb;
    const extremeDivergence = probabilityRange > 0.7; // Signals span >70% of range

    if (extremeDivergence) {
      console.warn(
        `[SignalFusion] Extreme signal divergence detected: range ${probabilityRange.toFixed(2)} (${minProb.toFixed(2)} to ${maxProb.toFixed(2)})`
      );
    }

    // Step 5: Assess data quality
    const dataQuality = assessDataQuality(agentSignals, state);

    // Step 6: Calculate fusion confidence
    let confidence = calculateFusionConfidence(
      agentSignals,
      weights,
      signalAlignment,
      dataQuality,
      config
    );

    // Reduce confidence significantly if extreme divergence
    if (extremeDivergence) {
      confidence = confidence * 0.5; // 50% penalty for extreme divergence
      console.warn(
        `[SignalFusion] Confidence reduced to ${confidence.toFixed(2)} due to extreme divergence`
      );
    }

    // Count MVP vs advanced agents
    const mvpAgents = ['market_microstructure', 'probability_baseline', 'risk_assessment'];
    const mvpAgentCount = agentSignals.filter((s) => mvpAgents.includes(s.agentName)).length;
    const advancedAgentCount = agentSignals.length - mvpAgentCount;

    // Create fused signal
    const fusedSignal = {
      fairProbability,
      confidence,
      signalAlignment,
      conflictingSignals,
      contributingAgents: activeAgents || agentSignals.map((s) => s.agentName),
      weights,
      metadata: {
        mvpAgentCount,
        advancedAgentCount,
        dataQuality,
        extremeDivergence,
        probabilityRange,
      },
    };

    return {
      fusedSignal,
      auditLog: [
        {
          stage: 'agent_signal_fusion',
          timestamp: Date.now(),
          data: {
            success: true,
            agentCount: agentSignals.length,
            mvpAgentCount,
            advancedAgentCount,
            fairProbability,
            confidence,
            signalAlignment,
            conflictCount: conflictingSignals.length,
            dataQuality,
            extremeDivergence,
            probabilityRange,
            weights,
            duration: Date.now() - startTime,
          },
        },
      ],
    };
  } catch (error) {
    console.error(
      '[SignalFusion] Signal fusion failed:',
      error instanceof Error ? error.message : String(error)
    );
    return {
      fusedSignal: null,
      auditLog: [
        {
          stage: 'agent_signal_fusion',
          timestamp: Date.now(),
          data: {
            success: false,
            error: error instanceof Error ? error.message : 'Unknown error',
            errorStack: error instanceof Error ? error.stack : undefined,
            duration: Date.now() - startTime,
          },
        },
      ],
    };
  }
}

/**
 * Create agent signal fusion node with bound configuration
 *
 * This factory function creates a node function that can be added to the LangGraph.
 *
 * @param config - Engine configuration
 * @returns Node function for LangGraph
 */
export function createAgentSignalFusionNode(config: EngineConfig) {
  return async (state: GraphStateType): Promise<Partial<GraphStateType>> => {
    return agentSignalFusionNode(state, config);
  };
}
